src/lib/Imp.hs

```diff
@@ -177,86 +223,76 @@ toImpOp (maybeDest, op) = case op of
-    emitSwitch cond' [emitStatement $ IInstr (IDo, liftInstr $ IThrowError), return ()]
+    emitSwitch cond' [emitStatement $ IInstr (Nothing, IThrowError), return ()]
     returnVal =<< intToIndex resultTy i'
     where (BaseTy tagBT) = TagRepTy
-  IdxSetSize n -> returnVal . toScalarAtom IdxRepTy =<< indexSetSize n
+  IdxSetSize n -> returnVal . toScalarAtom  =<< indexSetSize n
   IndexAsInt idx -> asInt $ case idx of
       Con (AnyValue t) -> anyValue t
       _                -> idx
     where
       asInt a = case a of
-        Con (Coerce (TC (IntRange   _ _  )) i) -> returnVal $ i
-        Con (Coerce (TC (IndexRange _ _ _)) i) -> returnVal $ i
-        _ -> returnVal . toScalarAtom IdxRepTy =<< indexToInt (getType idx) idx
+        Con (IntRangeVal   _ _   i) -> returnVal $ i
+        Con (IndexRangeVal _ _ _ i) -> returnVal $ i
+        _ -> returnVal . toScalarAtom =<< indexToInt idx
   Inject e -> do
-    let rt@(TC (IndexRange t low _)) = getType e
+    let (TC (IndexRange t low _)) = getType e
     offset <- case low of
-      InclusiveLim a -> indexToInt t a
-      ExclusiveLim a -> indexToInt t a >>= iaddI (fromScalarAtom $ IdxRepVal 1)
-      Unlimited      -> return (fromScalarAtom $ IdxRepVal 0)
-    restrictIdx <- indexToInt rt e
+      InclusiveLim a -> indexToInt a
+      ExclusiveLim a -> indexToInt a >>= iaddI (IIdxRepVal 1)
+      Unlimited      -> return (IIdxRepVal 0)
+    restrictIdx <- indexToInt e
     idx <- iaddI restrictIdx offset
     returnVal =<< intToIndex t idx
-  IndexRef ~(Var refVar@(_:>(RefTy h (Pi a)))) i -> do
-    refDest    <- getRef refVar
-    subrefDest <- destGet refDest i
-    subrefVar  <- freshVar (varName refVar :> RefTy h (snd $ applyAbs a i))
-    putRef subrefVar subrefDest
-    returnVal $ Var subrefVar
-  FstRef ~(Var refVar@(_:>(RefTy h (PairTy a _)))) -> do
-    ~(Dest (PairVal ref _)) <- getRef refVar
-    subrefVar <- freshVar (varName refVar :> RefTy h a)
-    putRef subrefVar $ Dest ref
-    returnVal $ Var subrefVar
-  SndRef ~(Var refVar@(_:>(RefTy h (PairTy _ b)))) -> do
-    ~(Dest (PairVal _ ref)) <- getRef refVar
-    subrefVar <- freshVar (varName refVar :> RefTy h b)
-    putRef subrefVar $ Dest ref
-    returnVal $ Var subrefVar
-  ArrayOffset arr idx off -> do
-    i <- indexToInt (getType idx) idx
-    let (TC (ArrayType t)) = getType arr
-    let resTy = tabTypeGet t i
-    arrSlice <- impOffset (fromArrayAtom arr) (fromScalarAtom off) resTy
-    returnVal $ toArrayAtom arrSlice
-  ArrayLoad arr -> returnVal . toScalarAtom resultTy =<< load (fromArrayAtom arr)
-  SliceOffset ~(Con (Coerce (TC (IndexSlice n l)) tileOffset)) idx -> do
-    i' <- indexToInt l idx
+  IndexRef refDest i -> returnVal =<< destGet refDest i
+  FstRef ~(Con (ConRef (PairCon ref _  ))) -> returnVal ref
+  SndRef ~(Con (ConRef (PairCon _   ref))) -> returnVal ref
+  PtrOffset arr off -> do
+    buf <- impOffset (fromScalarAtom arr) (fromScalarAtom off)
+    returnVal $ toScalarAtom buf
```

Comment:
nit: this could be a one liner

Reply:
With the `returnVal` combinator, the one-liner version becomes a bit too combinator-soup for me. I don't mind the extra lines here.